third_party/ibis/ibis_oracle/__init__.py

```diff
@@ -0,0 +1,81 @@
+# Copyright 2023 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import sqlalchemy as sa
+import re
+
+import ibis.expr.datatypes as dt
+from typing import Iterable, Literal, Tuple
+from ibis.backends.base.sql.alchemy import BaseAlchemyBackend
+from third_party.ibis.ibis_oracle.compiler import OracleCompiler
+from third_party.ibis.ibis_oracle.datatypes import _get_type
+
+
+class Backend(BaseAlchemyBackend):
+    name = "oracle"
+    compiler = OracleCompiler
+
+    def do_connect(
+        self,
+        host: str = "localhost",
+        user: str = None,
+        password: str = None,
+        port: int = 1521,
+        database: str = None,
+        protocol: str = "TCP",
+        url: str = None,
+        driver: Literal["cx_Oracle"] = "cx_Oracle",
+    ) -> None:
+        if url is None:
+            if driver != "cx_Oracle":
+                raise NotImplementedError(
+                    "cx_Oracle is currently the only supported driver"
+                )
+            dsn = """(description= (address=(protocol={})(host={})(port={}))
+            (connect_data=(service_name={})))""".format(
+                protocol, host, port, database
+            )
+            sa_url = sa.engine.url.URL.create(
+                "oracle+cx_oracle",
+                user,
+                password,
+                dsn,
+            )
+        else:
+            sa_url = sa.engine.url.make_url(url)
+        
+        self.database_name = sa_url.database
+        engine = sa.create_engine(sa_url, poolclass=sa.pool.StaticPool)
+
+        @sa.event.listens_for(engine, "connect")
+        def connect(dbapi_connection, connection_record):
+            with dbapi_connection.cursor() as cur:
+                cur.execute("ALTER SESSION SET TIME_ZONE='UTC'")
+    
+        super().do_connect(engine)
+
+
+    def _metadata(self, query)  -> Iterable[Tuple[str, dt.DataType]]:
+        if (
+            re.search(r"^\s*SELECT\s", query, flags=re.MULTILINE)
+            is not None
+        ):
+            query = f"({query})"
+
+        with self.begin() as con:
+            result = con.exec_driver_sql(f"SELECT * FROM {query} t0 WHERE ROWNUM <= 1")
```

Comment:
I think we are missing parentheses around the sub-query here. When testing custom-query I get:

```
sqlalchemy.exc.DatabaseError: (cx_Oracle.DatabaseError) ORA-00903: invalid table name
[SQL: SELECT * FROM select * from dvt_test.dvt_ora2pg_types t0 WHERE ROWNUM <= 1]
```

Reply:
Yeah - I think this is a result of me mistakenly deleting the re.IgnoreCase flag on [this line](https://github.com/GoogleCloudPlatform/professional-services-data-validator/pull/894/files/0ff3aaa42e769cb0a05e270cea3a85cae00dad90#diff-1b5e7ce512e4c118ff2bc62fff9cc2940a08c90e9a4affee9cc1f7c116449bfcR69) which would add the parentheses around the sub-query. Fixing this